jbmc/unit/java_bytecode/java_static_initializers/assignments_from_json.cpp

```diff
@@ -0,0 +1,412 @@
+  test_class_symbol.is_type = true;
+  test_class_symbol.type = [&] {
+    java_class_typet type;
+    // TODO java_class_typet constructors should ensure there is always a
+    // @class_identifier field, and that name is always set
+    type.set_name("java::" + class_name);
+    for(const auto &pair : components)
+      type.components().emplace_back(pair.first, pair.second);
+    return type;
+  }();
+  symbol_table.insert(test_class_symbol);
+}
+
+SCENARIO(
+  "assign_from_json",
+  "[core][java_static_initializers][assign_from_json]")
+{
+  symbol_tablet symbol_table;
+  const std::size_t max_user_array_length = 100;
+  std::unordered_map<std::string, object_creation_referencet> references;
+  std::unordered_multimap<irep_idt, symbolt> class_to_declared_symbols;
+
+  GIVEN("A class which has a static number in the provided JSON")
+  {
+    const json_objectt json = [] {
+      json_objectt entry{};
+      entry["field_name"] = [] {
+        json_objectt int_json;
+        int_json["value"] = json_numbert{"42"};
+        int_json["@type"] = json_stringt{"int"};
+        return int_json;
+      }();
+      entry["@type"] = json_stringt{"TestClass"};
+      return entry;
+    }();
+    class_to_declared_symbols.emplace("java::TestClass", [] {
+      symbolt field_symbol;
+      field_symbol.base_name = "field_name";
+      field_symbol.name = "field_name_for_codet";
+      field_symbol.type = java_int_type();
+      field_symbol.is_static_lifetime = true;
+      return field_symbol;
+    }());
+    add_clinit(symbol_table, "TestClass");
+    add_class_with_components(
+      symbol_table,
+      "TestClass",
+      {std::pair<std::string, typet>("@class_identifier", string_typet{}),
+       std::pair<std::string, typet>("field_name", java_int_type())});
+
+    const reference_typet test_class_type =
+      reference_type(struct_tag_typet{"java::TestClass"});
+    code_blockt block;
+    optionalt<ci_lazy_methods_neededt> option{};
+    assign_from_json(
+      symbol_exprt{"symbol_to_assign", test_class_type},
+      json,
+      "java::TestClass.<clinit>:()V",
+      block,
+      symbol_table,
+      option,
+      max_user_array_length,
+      references);
+    THEN("The instruction is the declaration of a symbol of TestClass* type")
+    {
+      const symbol_exprt declared_symbol =
+        make_query(block)[0].as<code_declt>()[0].as<symbol_exprt>().get();
+      REQUIRE(declared_symbol.type() == test_class_type);
+      THEN("The instruction allocates the symbol")
+      {
+        REQUIRE(
+          make_query(block)[1].as<code_assignt>()[0].as<symbol_exprt>().get() ==
+          declared_symbol);
+      }
+    }
+    THEN("The instruction assigns the symbol to `symbol_to_assign`")
+    {
+      REQUIRE(
+        make_query(block)[2]
+          .as<code_assignt>()[0]
+          .as<symbol_exprt>()
+          .get()
+          .get_identifier() == "symbol_to_assign");
+    }
+    THEN("The instruction zero-initializes the struct")
```

Comment:
:pick: The code here checks for "initializes the struct" but doesn't check for "zero".